caustic/parameter.py

```diff
@@ -20,92 +21,69 @@ class Parameter:
     def __init__(
         self, value: Optional[Union[Tensor, float]] = None, shape: Optional[tuple[int, ...]] = ()
     ):
-        """
-        Initializes an instance of the Parameter class.
-
-        Args:
-            value (Optional[Union[Tensor, float]]: The value of the parameter. Defaults to None.
-            shape (Optional[tuple[int, ...]]): The shape of the parameter. Defaults to an empty tuple.
-
-        Raises:
-            ValueError: If both value and shape are None, or if shape is provided and doesn't match the shape of the value.
-        """
         # Must assign one of value or shape
-        self._value = value
         if value is None:
             if shape is None:
                 raise ValueError("If value is None, a shape must be provided")
+            if not isinstance(shape, tuple):
+                raise TypeError("The shape of a parameter must be a tuple")
             self._shape = shape
         else:
-            value = torch.as_tensor(value)
+            value = torch.as_tensor(value).float() # set float32 as default dtype here
```

Comment:
no need to duplicate work, just do self.value = value and let the value setter handle the typing

Reply:
The init code is slightly different from the value setter, both because of this default, and also creating the private attributes _shape for the first time (setter never redefines the shape, which is important).

Reply:
In that case it seems strange to set the dtype in the init as well. maybe leave it as `value = torch.as_tensor(value)`?

Reply:
Agreed, I removed this default.